packages/aragon-api/src/index.js

```diff
@@ -228,19 +308,74 @@ export class AppProxy {
     // (because of the `Promise.resolve`).
     const wrappedReducer = (state, event) =>
       from(
-        Promise.resolve(reducer(state, event))
+        // Ensure a promise is returned even if the reducer returns an array
+        Promise.resolve(reducer(state, event)).catch(e => console.error(
+          `Error from app reducer:`, e)
+        )
       )
 
-    const store$ = initialState.pipe(
-      switchMap((initialState) =>
-        merge(
-          this.events(),
-          ...events
-        ).pipe(
-          mergeScan(wrappedReducer, initialState, 1),
-          map((state) => this.cache('state', state))
+    const getCurrentEvents = (fromBlock) => merge(
+      this.events(fromBlock),
+      ...externals.map(({ contract }) => contract.events(fromBlock))
+    )
+
+    // If `cachedFromBlock` is null there's no cache, `pastEvents` will use the initialisationBlock
+    // External contracts can specify their own `initializationBlock` which will be used in case the cache is empty
+    const getPastEvents = (cachedFromBlock, toBlock) => merge(
+      this.pastEvents(cachedFromBlock, toBlock),
+      ...externals.map(({ contract, initializationBlock }) => contract.pastEvents(cachedFromBlock || initializationBlock, toBlock))
+    )
+
+    const cachedState$ = this.state().pipe(first())
+    const cachedBlock$ = this.getCache(CACHED_BLOCK_KEY)
+    const latestBlock$ = this.web3Eth('getBlockNumber')
+    const initState$ = init ? from(init()) : from([null])
+
+    const store$ = combineLatest(cachedState$, initState$, cachedBlock$, latestBlock$).pipe(
+      switchMap(([cachedState, initState, cachedBlock, latestBlock]) => {
+        console.debug('- store - initState', initState)
+        console.debug('- store - cachedState', cachedState)
+        console.debug(`- store - cachedBlock ${cachedBlock} | latestBlock: ${latestBlock}`)
+
+        // The block up to which to fetch past events.
+        // The reduced state up to this point will be cached on every load
+        const pastEventsToBlock = latestBlock - BLOCK_REORG_MARGIN
+
+        console.debug(`- store - pastEvents: ${cachedBlock} -> ${pastEventsToBlock} (${pastEventsToBlock - cachedBlock} blocks)`)
+        console.debug(`- store - currentEvents$: from: ${pastEventsToBlock} -> future`)
+        // currentEvents$ is an observable for events starting from pastEventsToBlock
+        const currentEvents$ = getCurrentEvents(pastEventsToBlock)
+
+        const pastState$ = getPastEvents(cachedBlock, pastEventsToBlock).pipe(
+          // Prevent multiple subscriptions invoking duplicate calls
+          share(),
+          // single emission array of all pastEvents -> flatten to process events
+          flatMap(pastEvents => from(pastEvents)),
+          mergeScan(wrappedReducer, { ...cachedState, ...initState }, 1),
```

Comment:
This could be a place for further potential performance improvements on the first load.
The rough idea would be to change assumptions to handle side effects concurrently instead of blocking on every event's side effects.

A natural example is the vault which loads blocks to update the balances on every event. Now that the fetching of past events is a finite process, we can load all balances after reducing all past events. This would require another API change, perhaps in the form of an async callback that gets the state after reducing all past event.

Reply:
This is true, but I think it's already been unlocked with the `initialization` function, and doing this work immediately on initialization is actually better as it'd be much sooner than after the past events.

Here's a sketch for the vault:

1. Initialize: reload all balances of known tokens
1. Past events: in the `reducer()`, until it sees the special load trigger, avoid refreshing balances of already known tokens. If there are new tokens, fetch them
1. New events: refetch all required state in the `reducer()`

Reply:
True. This requires a small api change to pass the `cachedState` into the `init` function so that it can reload all balances for known tokens.

Reply:
Resolved in https://github.com/aragon/aragon.js/pull/297/commits/5f9e05387f1bf21bbce6102afdc60cea589fa5ff